### PR TITLE
A basic skeleton for OIDC

### LINKED ISSUE #2: Setup Swagger
## Description

Setup a swagger for OpenAPI Documentation of the [stencil-oidc-wrapper](https://github.com/techsavvyash/stencil-oidc-wrapper)

## Action Items

- [ ] Setup a swagger for OpenAPI Documentation of the [stencil-oidc-wrapper](https://github.com/techsavvyash/stencil-oidc-wrapper)

### PR DESCRIPTION
**Pull request by Zolo-Ryan**

### CODE REVIEW

**Comment by Zolo-Ryan** [comment]
Completed the api. IF any bugs encounterd please report.

File: prisma/migrations/20240525115351_init/migration.sql [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    "birthdate" TEXT NOT NULL,
+    "gender" TEXT NOT NULL,
+    "username" TEXT NOT NULL,
+    "email" TEXT NOT NULL,
+    "password" TEXT NOT NULL,
+    "tokens" TEXT
+);
+
+-- CreateIndex
+CREATE UNIQUE INDEX "User_username_key" ON "User"("username");
```
Comment (BackendBananeWala) [inline prisma/migrations/20240525115351_init/migration.sql:13]: In production user case we use **idx_user_username** as this provide much clarity and this is also a SQL naming conventions.

File: prisma/migrations/20240525115351_init/migration.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+-- CreateIndex
+CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
```
Comment (BackendBananeWala) [inline prisma/migrations/20240525115351_init/migration.sql:16]: Same as above.

File: src/api/api.controller.ts
Comment (BackendBananeWala) [inline src/api/api.controller.ts:1]: Module name should not be api

File: src/api/api.controller.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  constructor(
+    private readonly prismaService: PrismaService,
+    private readonly userService: UserService,
+    private readonly jwtService: JwtService,
+  ) {}
+  @Post('/login')
+  async loginRoute(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body('username') username: string,
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: Can we parse the body in dto? As we can use something as class-validator to validate inputs.

File: src/api/api.controller.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    @Body('password') password: string,
+    @Body('scopes') scopes: string,
+  ) {
+    //console.log('1');
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: Remove all the console logs after debug

File: src/api/api.controller.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    if (!jwtToken && (!username || !password)) {
+      res.status(400).send({
+        error: 'Invalid Credentials',
+        error_description: 'username, password missing in body',
+      });
+      return;
+    }
+    //console.log('2');
+    const user = jwtToken
+      ? await this.jwtService.verifyAsync(jwtToken, { secret: 'secret' })
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: can we save all the secure info in .env? we can also create something as .env.example for all the proper ENV_VARS in this file

File: src/api/api.controller.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+      return;
+    }
+    //console.log('4');
+
+    const finalScope = scopes ? scopes : 'openid';
+    //console.log(scopes);
+    let headersList = {
+      Accept: '*/*',
+      'Content-Type': 'application/x-www-form-urlencoded',
+      Authorization: 'Basic YXBwOmFfc2VjcmV0',
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: .env

File: src/api/api.controller.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    };
+
+    let bodyContent = `grant_type=client_credentials&scope=${finalScope}`;
+
+    let reqOptions = {
+      url: 'http://localhost:3000/oidc/token',
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: host_name in .env

File: src/api/api.controller.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      method: 'POST',
+      headers: headersList,
+      data: bodyContent,
+    };
+
+    let response = await axios.request(reqOptions);
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: As we are not reusing this variable can we change this to const ?

File: src/api/api.controller.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+      response.data.access_token,
+    );
+
+    if (!jwtToken) {
+      const token = await this.jwtService.signAsync(
+        {
+          id: user.id,
+          sub: response.data.access_token,
+        },
+        { secret: 'secret' },
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: .env

File: src/api/api.controller.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        error: 'No token given',
+        error_description:
+          "User with the given username and password doesn't exist",
+      });
+      return;
+    }
+    let headersList = {
+      Accept: '*/*',
+      'Content-Type': 'application/x-www-form-urlencoded',
+      Authorization: 'Basic YXBwOmFfc2VjcmV0',
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: .env

File: src/api/api.controller.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    //console.log(response.data);
+    return res.send(response.data);
+  }
+
+  @Post('/signup')
+  async signupRoute(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body()
+    body: {
```
Comment (BackendBananeWala) [inline src/api/api.controller.ts]: dto ?

File: src/app.controller.ts [hunk 12]
```diff
@@ -1,12 +1,89 @@
...
+    @Req() req: Request,
+    @Res() res: Response,
+    @Query('error') error: string,
+    @Query('error_description') error_description: string,
+  ) {
+    const result: Record<string, any> = {
+      query: { error, error_description },
+      accountId: null,
+      scopes: null,
+      origin: 'http://localhost:3000',
```
Comment (BackendBananeWala) [inline src/app.controller.ts]: .env

File: src/app.controller.ts [hunk 13]
```diff
@@ -1,12 +1,89 @@
...
+          code,
+        }),
+        {
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            Authorization: 'Basic YXBwOmFfc2VjcmV0',
+          },
+        },
+      );
+      console.log('THis is result data', result.data);
```
Comment (BackendBananeWala) [inline src/app.controller.ts]: remove after debug

File: src/app.controller.ts [hunk 14]
```diff
@@ -1,12 +1,89 @@
...
+        qs.stringify({
+          client_id: 'app',
+          grant_type: 'authorization_code',
+          redirect_uris: 'http://localhost:3000/callback',
+          code,
+        }),
+        {
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            Authorization: 'Basic YXBwOmFfc2VjcmV0',
```
Comment (BackendBananeWala) [inline src/app.controller.ts]: .env

File: src/user/user.controller.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    constructor(private readonly userService: UserService){}
+
+    @Get('/all')
+    async getAllUsers(){
+        return this.userService.allUsers();
+    }
+
+    @Post('/signup')
+    async register(
+        @Body() userData: {birthdate: string, gender: string, username: string, email: string, password: string,tokens?: string}
```
Comment (BackendBananeWala) [inline src/user/user.controller.ts]: dto

**Review by BackendBananeWala — COMMENTED** [review]
Iteration : #1

File: prisma/migrations/20240525115351_init/migration.sql (hunk 1 above)
Comment (Zolo-Ryan) [reply prisma/migrations/20240525115351_init/migration.sql:13]: This is auto generated by prisma.

File: prisma/migrations/20240525115351_init/migration.sql (hunk 2 above)
Comment (Zolo-Ryan) [reply prisma/migrations/20240525115351_init/migration.sql:16]: This is auto generated by prisma.

File: package.json [hunk 16]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "stencil-oidc-wrapper",
+  "name": "nest-oidc-provider-final",
```
Comment (techsavvyash) [inline package.json]: Revert this

File: .gitignore [hunk 17]
```diff
@@ -1,6 +1,8 @@
 # compiled output
 /dist
 /node_modules
+.env 
+package-lock.json
```
Comment (techsavvyash) [inline .gitignore]: remove this out of `.gitignore`

File: src/app.controller.ts [hunk 18]
```diff
@@ -1,12 +1,198 @@
...
+
+import { PrismaService } from 'src/prisma/prisma.service';
+import { UserService } from 'src/user/user.service';
+import { JwtService } from '@nestjs/jwt';
+import { CreateUserDto, LoginDTO } from './dto/user.dto';
 
 @Controller()
 export class AppController {
-  constructor(private readonly appService: AppService) {}
+  private readonly logger = new Logger(AppController.name);
```
Comment (techsavvyash) [inline src/app.controller.ts]:
```suggestion
  private readonly logger: Logger;
  constructor(...) {
   this.logger = new Logger();
  }
```

File: src/app.controller.ts [hunk 19]
```diff
@@ -1,12 +1,198 @@
...
+  constructor(
+    private readonly appService: AppService,
+    private readonly userService: UserService,
+    private readonly prismaService: PrismaService,
+    private readonly jwtService: JwtService,
+  ) {}
+
+  @Get('/')
```
Comment (techsavvyash) [inline src/app.controller.ts]: change this to a `/health` and return the response as 
```json
{
   "status": "live"
}
```

File: src/app.controller.ts [hunk 20]
```diff
@@ -1,12 +1,198 @@
...
+  @Post('/login')
+  async loginRoute(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body() body: LoginDTO,
+  ) {
+    const { username, password, scopes, resource, grant_type } = body;
+    const jwtToken = req.cookies?.jwt;
+    if (!jwtToken && (!username || !password)) {
+      res.status(400).send({
```
Comment (techsavvyash) [inline src/app.controller.ts]: raise an expection instead of doing like this:
example snippet:
```suggestion
throw new InvalidRequestException({
        error: 'Invalid Credentials',
        error_description: 'username, password missing in body',
});
```

File: src/app.controller.ts [hunk 21]
```diff
@@ -1,12 +1,198 @@
...
+        error: 'Invalid Credentials',
+        error_description: 'username, password missing in body',
+      });
+      return;
+    }
+    if (resource && !resource.includes(':')) {
+      res.status(401).send({
```
Comment (techsavvyash) [inline src/app.controller.ts]: similar to above

File: src/app.controller.ts [hunk 22]
```diff
@@ -1,12 +1,198 @@
...
+    };
+  }
+
+  @Post('/login')
+  async loginRoute(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body() body: LoginDTO,
+  ) {
+    const { username, password, scopes, resource, grant_type } = body;
```
Comment (techsavvyash) [inline src/app.controller.ts:164]: also abstract out the entire logic in a service file.

File: src/app.controller.ts [hunk 23]
```diff
@@ -1,12 +1,198 @@
...
+      message: resource ? 'jwt token generated': 'opaque token generated',
+      data: response.data
+    });
+  }
+
+  // for verifying opaque tokens
+  @Post('/opaque-verify')
+  async opaque_verify(@Res() res: Response, @Body('token') token: string) {
+    if (!token) {
+      res.status(401).send({
```
Comment (techsavvyash) [inline src/app.controller.ts]: throw an exception here

File: src/app.controller.ts [hunk 24]
```diff
@@ -1,12 +1,198 @@
...
+    return res.send({
+      message: resource ? 'jwt token generated': 'opaque token generated',
+      data: response.data
+    });
+  }
+
+  // for verifying opaque tokens
+  @Post('/opaque-verify')
+  async opaque_verify(@Res() res: Response, @Body('token') token: string) {
+    if (!token) {
```
Comment (techsavvyash) [inline src/app.controller.ts:338]: abstract the logic out in a service function.

File: src/app.controller.ts [hunk 25]
```diff
@@ -1,12 +1,198 @@
...
+    };
+    const response = await axios.request(reqOptions);
+
+    return res.send(response.data);
+  }
+
+  // to verify jwt created tokens, whose public keys are available
+  @Post('/jwt-verify')
+  async jwt_verify(@Body('token') token: string){
+    if (!token) {
```
Comment (techsavvyash) [inline src/app.controller.ts:490]: abstract the logic out in a service function.

File: src/app.controller.ts [hunk 26]
```diff
@@ -1,12 +1,198 @@
...
+
+    return val;
+  }
 
-  @Get()
-  getHello(): string {
-    return this.appService.getHello();
+  @Post('/signup')
+  async signupRoute(
+    @Req() req: Request,
```
Comment (techsavvyash) [inline src/app.controller.ts]: remove req, res

File: src/app.controller.ts [hunk 27]
```diff
@@ -1,12 +1,198 @@
...
+    @Res() res: Response,
+    @Body()
+    body: CreateUserDto,
+  ) {
+    const { username, password, gender, birthdate, email } = body;
+    if (!username || !password || !gender || !birthdate || !email) {
+      res.status(400).send({
```
Comment (techsavvyash) [inline src/app.controller.ts]: throw an exception

File: src/app.controller.ts [hunk 28]
```diff
@@ -1,12 +1,198 @@
...
-    return this.appService.getHello();
+  @Post('/signup')
+  async signupRoute(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body()
+    body: CreateUserDto,
+  ) {
+    const { username, password, gender, birthdate, email } = body;
+    if (!username || !password || !gender || !birthdate || !email) {
```
Comment (techsavvyash) [inline src/app.controller.ts:646]: abstract the logic out in a service.

File: src/app.module.ts [hunk 29]
```diff
@@ -1,10 +1,20 @@
...
+import { PrismaModule } from './prisma/prisma.module';
+import { UserService } from './user/user.service';
+import { PrismaService } from './prisma/prisma.service';
+import { JwtModule } from '@nestjs/jwt';
 
 @Module({
-  imports: [],
+  imports: [OidcModule, UserModule, PrismaModule,JwtModule.register({
+    global: true,
+    secret: "secret",
```
Comment (techsavvyash) [inline src/app.module.ts]: move to an env var.

File: src/app.module.ts [hunk 30]
```diff
@@ -1,10 +1,20 @@
...
+    signOptions: { expiresIn: '1hr' },
```
Comment (techsavvyash) [inline src/app.module.ts]: move to an env var.

File: src/oidc/oidc.controller.ts [hunk 31]
```diff
@@ -0,0 +1,18 @@
+import { All, Controller, Req, Res } from '@nestjs/common';
+import { Request, Response } from 'express';
+import getCallbackFunction from './oidc.util';
+
+// const Provider = getProvider();
```
Comment (techsavvyash) [inline src/oidc/oidc.controller.ts]: remove redundant comment

File: src/oidc/oidc.util.ts [hunk 32]
```diff
@@ -0,0 +1,190 @@
+import { Configuration } from 'oidc-provider';
+
+let Provider, oidc;
+const configuration: Configuration = {
+  // ... see the available options in Configuration options section
+  features: {
+    // claimsParameter:{
+    //     enabled: false
+    // },
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: remove redundant comments.

File: src/oidc/oidc.util.ts [hunk 33]
```diff
@@ -0,0 +1,190 @@
...
+             audience: 'urn:api',
+             accessTokenTTL: 1 * 60 * 60, // 1 hour
+             accessTokenFormat: 'jwt',
+           };
+         }
+
+         throw new Error('Invalid target');
+       },
+     }, */
+    /* To turn off default pages and routes */
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: unless any of these comments are for documentation purposes, remove them

File: src/oidc/oidc.util.ts [hunk 34]
```diff
@@ -0,0 +1,190 @@
...
+      return `/interaction/${interaction.uid}`;
+    },
+  },
+  clientBasedCORS(ctx, origin, client) {
+    return false;
+  },
+
+  //   claims:{
+  //     profile: ['username','gender','birthdate','email'],
+  //   },
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: remove unless this is for document purposes, if this is for document purposes, change this to use JSDoc format.

File: src/oidc/oidc.util.ts [hunk 35]
```diff
@@ -0,0 +1,190 @@
...
+
+  //   }
+
+  //   interactions: {
+  //     url(ctx, interaction) {
+  //       // eslint-disable-line no-unused-vars
+  //       return `/interaction/${interaction.uid}`;
+  //     },
+  //   }, */
+
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: remove unless this is for document purposes, if this is for document purposes, change this to use JSDoc format.

File: src/oidc/oidc.util.ts [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+  // own cookies.key needed rather than default
+  cookies: {
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: why are we using cookies? either remove them or document proper reason to be using them.

File: src/oidc/oidc.util.ts [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+  //       'website',
+  //       'zoneinfo',
+  //     ],
+  //   },
+  //   features: {
+  //     devInteractions: { enabled: false }, // defaults to true
+
+  //     deviceFlow: { enabled: true }, // defaults to false
+  //     revocation: { enabled: true }, // defaults to false
+  //   }, */
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: remove unless this is for document purposes, if this is for document purposes, change this to use JSDoc format.

File: src/oidc/oidc.util.ts [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+        crv: 'P-256',
+        d: 'K9xfPv773dZR22TVUB80xouzdF7qCg5cWjPjkHyv7Ws',
+        kty: 'EC',
+        use: 'sig',
+        x: 'FWZ9rSkLt6Dx9E3pxLybhdM6xgR5obGsj5_pqmnz5J4',
+        y: '_n8G69C-A2Xl4xUW2lF0i8ZGZnk_KPYrhv4GbTGu5G4',
+      },
+    ],
+  },
+};
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: why is this here? ideally this should be persisted in the db and fetched from there.

File: src/oidc/oidc.util.ts [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+    `${process.env.HOST_NAME}:${process.env.HOST_PORT}`,
+    configuration,
+  );
+});
+async function getCallbackFunction() {
+  return oidc.callback();
+}
+
+export default getCallbackFunction;
+// export { Provider, oidc };
```
Comment (techsavvyash) [inline src/oidc/oidc.util.ts]: remove

File: src/prisma/prisma.service.spec.ts
Comment (techsavvyash) [inline src/prisma/prisma.service.spec.ts:1]: remove this file.

**Review by techsavvyash — CHANGES_REQUESTED** [review]
